hazelcast/test/src/HazelcastTests8.cpp

```diff
@@ -2344,6 +2344,87 @@ TEST_F(
 } // namespace client
 } // namespace hazelcast
 
+namespace hazelcast {
+namespace client {
+namespace test {
+namespace thread_pool {
+
+class ThreadPoolTest
+  : public ClientTest
+  , public testing::WithParamInterface<int32_t>
+{
+protected:
+    struct ThreadState
+    {
+        explicit ThreadState(int latch_count)
+          : latch1(latch_count)
+        {
+        }
+
+        boost::latch latch1;
+        std::set<boost::thread::id> thread_ids;
+    };
+
+protected:
+    static void SetUpTestCase()
+    {
+        instance = new HazelcastServer(default_server_factory());
+    }
+
+    static void TearDownTestCase()
+    {
+        delete client;
+        delete instance;
+
+        client = nullptr;
+        instance = nullptr;
+    }
+
+    static HazelcastServer* instance;
+    static hazelcast_client* client;
+};
+
+HazelcastServer* ThreadPoolTest::instance = nullptr;
+hazelcast_client* ThreadPoolTest::client = nullptr;
+
+TEST_P(ThreadPoolTest, testEqualThreadAndJobs)
+{
+    int32_t num_of_thread = 5;
+    int32_t num_of_jobs = GetParam();
+    client_config config;
+    config.set_executor_pool_size(num_of_thread);
+
+    client = new hazelcast_client{ new_client(std::move(config)).get() };
+
+    spi::ClientContext ctx(*client);
+    auto state = std::make_shared<ThreadState>(num_of_jobs);
+    std::mutex mutex_for_thread_id;
+
+    ASSERT_EQ(0, state->thread_ids.size());
+    for (int i = 0; i < num_of_jobs; i++) {
+        ctx.get_client_execution_service().get_user_executor().submit(
+          [state, &mutex_for_thread_id]() {
+              std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
why sleep?

Reply:
to be sure that the job takes some time. It may go to the same thread if the job instantly returns.

Reply:
This is still time sensitive and may fail on GA runs depending on machine speed. Can we make it more time insensitive? E.g. inside the thread we can have all threads wait on a latch and sync each other and then continue.

Reply:
I removed the sleep method and used a barrier to sync the threads.